### PR TITLE
Enable test connection button

### PR DESCRIPTION
**Pull request by takirala**
Summary:
Test connection to Mesos via `Test Connection` button in form.

JIRA : DCOS_OSS-5079

Co-Authored-By: Karsten Jeschkies <kjeschkies@mesosphere.com>

### CODE REVIEW

File: src/main/java/org/jenkinsci/plugins/mesos/MesosCloud.java [hunk 1]
```diff
@@ -155,6 +155,11 @@ public boolean canProvision(Label label) {
         .toCompletableFuture();
   }
 
+  @Override
+  public DescriptorImpl getDescriptor() {
+    return (DescriptorImpl) super.getDescriptor();
```
Comment (jeschkies) [inline src/main/java/org/jenkinsci/plugins/mesos/MesosCloud.java]: Just out of curiosity. Why was this overriden?

File: src/main/java/org/jenkinsci/plugins/mesos/MesosCloud.java [hunk 2]
```diff
@@ -167,12 +172,36 @@ public String getDisplayName() {
       return "Mesos Cloud";
     }
 
-    // TODO: validate URLs
+    /** Test connection from Mesos Cloud configuration page. */
+    public FormValidation doTestConnection(@QueryParameter("mesosMasterUrl") String mesosMasterUrl) {
+      mesosMasterUrl = mesosMasterUrl.trim();
+
+      if (mesosMasterUrl.startsWith("zk://")) {
+        return FormValidation.warning("Zookeeper paths can be used, but the connection cannot be " +
```
Comment (jeschkies) [inline src/main/java/org/jenkinsci/plugins/mesos/MesosCloud.java]: This should be done by the actual validation code see https://jira.mesosphere.com/browse/DCOS_OSS-5046.

File: src/main/java/org/jenkinsci/plugins/mesos/MesosCloud.java [hunk 3]
```diff
@@ -167,12 +172,36 @@ public String getDisplayName() {
...
+                "tested prior to saving this page.");
+      }
 
-    /** Test connection from configuration page. */
-    public FormValidation doTestConnection(
-        @QueryParameter("mesosMasterUrl") String mesosMasterUrl) {
-      throw new NotImplementedException("Connection testing is not supported yet.");
+      if (mesosMasterUrl.startsWith("http://") || mesosMasterUrl.startsWith("https://") ) {
+        return FormValidation.error("Please omit 'http(s)://'.");
```
Comment (jeschkies) [inline src/main/java/org/jenkinsci/plugins/mesos/MesosCloud.java]: See above. Also, the Mesos master URL should be a valid `java.net.URL` or `java.net.URI`.

File: src/main/java/org/jenkinsci/plugins/mesos/MesosCloud.java [hunk 4]
```diff
@@ -167,12 +172,36 @@ public String getDisplayName() {
...
+      }
+
+      try {
+        // URL requires the protocol to be explicitly specified.
+        HttpURLConnection urlConn =
+                (HttpURLConnection) new URL("http://" + mesosMasterUrl).openConnection();
```
Comment (jeschkies) [inline src/main/java/org/jenkinsci/plugins/mesos/MesosCloud.java]: `mesosMasterUrl` should already include `http`. See my comment above.

File: src/main/java/org/jenkinsci/plugins/mesos/MesosCloud.java [hunk 5]
```diff
@@ -167,12 +172,36 @@ public String getDisplayName() {
...
+        urlConn.connect();
+        int code = urlConn.getResponseCode();
+        urlConn.disconnect();
+
+        if (code == 200) {
+          return FormValidation.ok("Connected to Mesos successfully");
```
Comment (jeschkies) [inline src/main/java/org/jenkinsci/plugins/mesos/MesosCloud.java]:
```suggestion
          return FormValidation.ok("Mesos ping succeeded.");
```

**Review by jeschkies — DISMISSED** [review]
I wonder if we should use the Mesos Client from USI and also which URL we want to ping.

File: src/main/java/org/jenkinsci/plugins/mesos/MesosCloud.java (hunk 1 above)
Comment (jeschkies) [reply src/main/java/org/jenkinsci/plugins/mesos/MesosCloud.java]: Ping @takirala

File: src/main/java/org/jenkinsci/plugins/mesos/MesosCloud.java (hunk 3 above)
Comment (jeschkies) [reply src/main/java/org/jenkinsci/plugins/mesos/MesosCloud.java]: Please checkout https://github.com/mesosphere/mesos-plugin/pull/38. `isValidUrl` should suffice.

File: src/main/java/org/jenkinsci/plugins/mesos/MesosCloud.java [hunk 6]
```diff
@@ -167,12 +172,38 @@ public String getDisplayName() {
...
+      }
+
+      if (mesosMasterUrl.startsWith("http://") || mesosMasterUrl.startsWith("https://")) {
+        return FormValidation.error("Please omit 'http(s)://'.");
+      }
+
+      try {
+        // URL requires the protocol to be explicitly specified.
+        HttpURLConnection urlConn =
+            (HttpURLConnection) new URL("http://" + mesosMasterUrl).openConnection();
```
Comment (jeschkies) [inline src/main/java/org/jenkinsci/plugins/mesos/MesosCloud.java]: I think we should follow redirects.

**Review by jeschkies — CHANGES_REQUESTED** [review]
Thanks for your effort. Could you add an integration test?

**Review by jeschkies — APPROVED** [review]

**Comment by takirala** [comment]
Thanks @jeschkies

File: src/main/java/org/jenkinsci/plugins/mesos/MesosCloud.java [hunk 7]
```diff
@@ -263,10 +265,40 @@ public FormValidation doCheckJenkinsUrl(@QueryParameter String jenkinsUrl) {
...
+    /**
+     * Test connection from configuration page.
+     *
+     * @param mesosMasterUrl The Mesos master URL set by the user.
+     * @return Whether the URL is correct and reachable or a validation error.
+     */
     public FormValidation doTestConnection(
         @QueryParameter("mesosMasterUrl") String mesosMasterUrl) {
-      throw new NotImplementedException("Connection testing is not supported yet.");
+      FormValidation urlValidation = doCheckMesosMasterUrl(mesosMasterUrl);
```
Comment (kvish) [inline src/main/java/org/jenkinsci/plugins/mesos/MesosCloud.java:276]: does this have to be manually called?

https://github.com/mesosphere/jenkins-mesos-plugin/blob/master/src/main/java/org/jenkinsci/plugins/mesos/MesosCloud.java#L188

would this validate the Mesos master url automatically when the  test connection button is pressed or only if the form is submitted?

Comment (jeschkies) [reply src/main/java/org/jenkinsci/plugins/mesos/MesosCloud.java:276]: I don't think it is called automatically. Let me try.

Comment (jeschkies) [reply src/main/java/org/jenkinsci/plugins/mesos/MesosCloud.java:276]: ..I checked. The validation is not called. However, the try catch will catch the `java.net.MalformedURLException`. I'd keep the validation.

**Review by kvish — APPROVED** [review]

**Pull request merged**